Update for ProtonVPN update-port.sh

BigRedBrent opened the pull request:
Disabled exiting on errors to allow the script to keep running even if commands fail, so that it can retry after 45 seconds on the next loop. (I've let the script run for several days, and it looks like this has resolved the issue of it stopping prematurely.)

Returned functionality to allow and deny ports in the firewall as necessary.

Modified the bind_trans function to be less complicated. Removed multiple attempts per loop, because it can retry after 45 seconds on the next loop.

Will now double check to make sure the port has actually changed on the next 45 second loop. This seems to catch incorrect port changes, or at the very least does not switch to a very short lived port change.

<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Leave this section emtpy if this PR is NOT a breaking change.
-->
```txt
<placeholder>
```


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
```txt
<placeholder>
```


## Type of change
<!--
  What type of change does your PR introduce?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New provider (thank you!)
- [ ] Updated provider (thank you!)
- [ ] New feature (which adds functionality to a provider script/repo usage)
- [ ] Breaking change (fix/feature causing existing functionality to break)


## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: ```fixes #```
- This PR is related to issue: ```relates to #```
- Link to documentation updated (if done separately): ```https://...```

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [ ] The code change is tested and works locally.
- [ ] There is no commented out code in this PR.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated

<!-- Please check *Preview* before submitting -->

BigRedBrent committed:
Update update-port.sh

Removed multiple attempts per loop, as they are unnecessary, since it can just retry the attempt on the next loop. This makes the code a bit less complicated.

BigRedBrent committed:
Update update-port.sh

BigRedBrent committed:
Update update-port.sh

Disabled exiting on errors to allow the script to keep running even if commands fail

BigRedBrent committed:
Update update-port.sh

Removed echo_debug function

BigRedBrent committed:
Update update-port.sh

BigRedBrent committed:
Update update-port.sh

BigRedBrent committed:
Update update-port.sh

BigRedBrent committed:
Update update-port.sh

BigRedBrent committed:
Update update-port.sh

Added functionality for firewall

BigRedBrent committed:
Update update-port.sh

BigRedBrent committed:
Update update-port.sh

BigRedBrent committed:
Update update-port.sh

BigRedBrent committed:
Update update-port.sh

BigRedBrent committed:
Update update-port.sh

BigRedBrent committed:
Update update-port.sh

BigRedBrent committed:
Update update-port.sh

Renamed port variables to be clearer to understand

BigRedBrent committed:
Update update-port.sh

BigRedBrent committed:
Update update-port.sh

BigRedBrent committed:
Update update-port.sh

BigRedBrent committed:
Update update-port.sh

BigRedBrent committed:
Update update-port.sh

BigRedBrent committed:
Update update-port.sh

Replaced test with brackets

BigRedBrent committed:
Update update-port.sh

BigRedBrent committed:
Update update-port.sh

BigRedBrent committed:
Update update-port.sh

BigRedBrent committed:
Update update-port.sh

BigRedBrent committed:
Update update-port.sh

BigRedBrent committed:
Merge branch 'main' into patch-3

BigRedBrent commented:
@oregonpillow

I was unable to get the firewall to work also, but I was returning the functionality as it was originally there before my first edit that I made, in the hopes that someone will figure out a way to repair it.

The other modifications I made to this commit seem to work pretty well with the firewall disabled. Much better than my first edit did.

BigRedBrent commented:
Also, as long as a kill switch has been set, I see no reason for port enabling and port blocking. Has a kill switch been enabled in this docker container?

If a kill switch hasn't been added, I used AI to whip this up:
```

# Manually set to enable the kill switch
ENABLE_KILL_SWITCH="true"

init_kill_switch() {
    if [[ "$ENABLE_KILL_SWITCH" != "true" ]]; then
        return 0
    fi
    echo -e "$(date '+%Y-%m-%d %T')\tInitializing VPN kill switch"

    # Flush existing rules
    iptables -F
    iptables -X

    # Default policies: drop everything
    iptables -P INPUT DROP
    iptables -P OUTPUT DROP
    iptables -P FORWARD DROP

    # Allow loopback traffic
    iptables -A INPUT -i lo -j ACCEPT
    iptables -A OUTPUT -o lo -j ACCEPT

    # Allow established and related connections
    iptables -A INPUT -m conntrack --ctstate ESTABLISHED,RELATED -j ACCEPT
    iptables -A OUTPUT -m conntrack --ctstate ESTABLISHED,RELATED -j ACCEPT

    # Allow all traffic over any VPN interface
    for iface in $(ip -o link show | awk -F': ' '{print $2}' | grep -E '^(tun|tap|wg|ppp)'); do
        iptables -A INPUT -i "$iface" -j ACCEPT
        iptables -A OUTPUT -o "$iface" -j ACCEPT
    done

    # Allow local network for web UI if set
    if [[ -n "${LOCAL_NETWORK:-}" ]]; then
        iptables -A INPUT -s "${LOCAL_NETWORK}" -j ACCEPT
        iptables -A OUTPUT -d "${LOCAL_NETWORK}" -j ACCEPT
    fi

    # Allow Transmission web UI port (default 9091)
    iptables -A INPUT -p tcp --dport "${TRANSMISSION_RPC_PORT:-9091}" -j ACCEPT

    echo -e "$(date '+%Y-%m-%d %T')\tVPN kill switch initialized"
}

init_kill_switch

```

BigRedBrent committed:
Update update-port.sh

Added a simple initialization echo message

BigRedBrent committed:
Update update-port.sh

Latest image does not have natpmpc installed by default. Added script to automatically attempt to install natpmpc if it is not already installed.

BigRedBrent committed:
Update update-port.sh

oregonpillow commented:
Just as an FYI, I recently switched to gluetun + transmission as an alternative to Haugene. I've been using Haugene for years (4-5?) but have to say that the gluetun stack works much better for me. With proton, the port forwarding works immediately.

If you're not set on Haugene I just want to let you know it exists. But if you need an AIO solution, then Haugene is the way to go.

BigRedBrent committed:
Update update-port.sh

Added install_package function to install and update packages as necessary.

BigRedBrent committed:
Update update-port.sh

BigRedBrent commented:
I couldn't get gluetun + transmission to work. Perhaps it's because I'm using a custom QNAP network configuration.

BigRedBrent committed:
Update update-port.sh

Now does a health check before starting the script and no longer updates already installed packages.

BigRedBrent committed:
Update update-port.sh

BigRedBrent committed:
Update update-port.sh

BigRedBrent committed:
Update update-port.sh

Added update-port: prefix to echo commands

BigRedBrent committed:
Update update-port.sh

added a tab to echo commands to make them visually easier to read

BigRedBrent committed:
Update update-port.sh

replaced echo command with log function

BigRedBrent committed:
Update update-port.sh

BigRedBrent committed:
Update update-port.sh

BigRedBrent committed:
Update update-port.sh

BigRedBrent committed:
Fix port binding for Transmission 4.1.0's JSON-RPC 2.0 and snake_case RPC changes

Fixes port binding on Transmission 4.1.0, which switched to JSON-RPC 2.0 and snake_case keys — .result is now the payload object instead of "success", and peer-port is now peer_port. Adds two helpers, rpc_ok() and session_port(), that accept either response shape, so it works on both 4.0.x and 4.1.x with no version detection.

haugene commented:
Hey @BigRedBrent! Sorry about the delay on reviews.

Thank you for fixing this :raised_hands: I'm ready to merge, but just wanted to check a couple of things with you as I don't know this code line by line.

I'm curious about the `double_check` logic, has it been returning alternating ports? It's a bit tough to read and I wonder if it could be simplified/removed.

The UFW stuff. Not the best maintained feature in any parts of the image TBH, but after a port change `current_port != last_port` will be the steady state? Then it will run `ufw deny` on every loop, and that doesn't remove the previous `ufw allow` rule. Should it be `ufw delete allow`?

You've removed retries for the `bind_trans` function. But switching from `sleep` to actually checking the healthcheck script might do the trick and remove RPC flakyness?

And. This script installs packages into the image, possibly. A bit surprising for a config repo, but maybe it's worth it compared to just logging a clear error message that would need to be raised as an issue. The image ships all three right now, so it's not needed for new versions of the image - but maybe for backwards compatibility? Could be an opt-in.

Final note. I'll merge it as-is if you want. The config repo was made to separate some concerns and ownership of code and I think the users of ProtonVPN owns this code more than me. And if someone want stable UFW functionality then that's up to them. I just wanted to make a note and see if you had any thoughts.

Let me know :smile:

BigRedBrent committed:
Attempted to fix the firewall updater

Attempted to fix the firewall updater

BigRedBrent committed:
Simplify rpc_ok function response handling

Refactor rpc_ok function to clarify response shape handling.

BigRedBrent committed:
Refactor firewall rule management in update-port.sh

BigRedBrent committed:
Refactor ufw rule checks to inline commands

BigRedBrent committed:
Fix port update logic in update-port.sh
